rfc-141-application-healthchecks.md

```diff
@@ -0,0 +1,224 @@
+# Application Healthchecks
+
+## Summary
+
+We currently use application healthchecks for three things:
+
+- For load balancing, to determine if an instance is ready to serve
+  requests.
+- For continuous deployments, to determine (along with other smoke
+  tests) if a release can be automatically pushed to production.
+- For alerting, so that 2ndline know if an instance has a problem
+  which needs manual intervention to fix.
+
+We implement healthchecks using the [`GovukHealthcheck`][] module in
+[govuk_app_config][].
+
+Currently, a healthcheck response is always served with an HTTP status
+of 200, and indicates in the body whether the instance's status is
+"ok", "warning", or "critical".  However, AWS load balancers use the
+HTTP status, not the response body, to determine whether an instance
+is healthy.  So we will continue to send requests to an instance which
+is in a "warning" or "critical" state.
+
+This RFC proposes that we standardise on our healthchecks returning a
+200 or a 500 status code, which has some further ramifications on
+monitoring and alerting (discussed below).
+
+A full list of healthchecks which will need updating is given at the
+end of the document.
+
+## Problem
+
+We overload the meaning of "healthcheck", and cover both *application
+health* as well as *instance health* with our `/healthcheck`
+endpoints.  This introduces confusion, and means that we cannot
+currently use healthchecks in load balancing.
+
+Here are some examples of healthchecks which are suitable for load
+balancing and for post-deployment checks:
+
+- [`GovukHealthcheck::ActiveRecord`][]
+- [`GovukHealthcheck::Mongoid`][]
+- [`GovukHealthcheck::RailsCache`][]
+- [`GovukHealthcheck::Redis`][]
+- [`GovukHealthcheck::SidekiqRedis`][]
+
+If an app can't talk to a backing service it relies on, then it
+probably can't be trusted to handle any requests.  One of these
+healthchecks failing could indicate a networking or a configuration
+issue.
+
+Here are some examples of healthchecks which are unsuitable for load
+balancing or for post-deployment checks:
+
+- [`GovukHealthcheck::SidekiqQueueCheck`][], this is an abstract check
+  which other checks extend.
+  [`GovukHealthcheck::SidekiqQueueLatencyCheck`][] is an instance of
+  this check.
+- [`GovukHealthcheck::ThresholdCheck`][], this is another abstract
+  check which other checks extend.
+  [`GovukHealthcheck::SidekiqRetrySizeCheck`][] is an instance of this
+  check.
+- [`Healthcheck::ApiTokens`][] in [signon][], which reports that an
+  API token needs rotating.
+
+If one of these healthchecks fail, the app instances themselves are
+probably fine.  The problem is elsewhere.
+
+Two of these indicate capacity problems outside of the instance, and
+the third is an alert that a manual maintenance procedure needs to be
+performed some time in the next two months.  They all share the
+property that if one instance reports a failure, all will.  This makes
+them unsuitable for load balancing purposes.
```

Comment:
It sounds quite reasonable to remove these, they seemed to make more sense when Icinga was pretty much the only consumer.

It does seem to leave a bit of a gap about how application state can be pulled for monitoring. I guess we were using them as a rather crude method for continuous pulling of state, the alternative approach (in Email Alert API we run a [Sidekiq worker to continuously push metrics to graphite](https://github.com/alphagov/email-alert-api/blob/master/app/workers/metrics_collection_worker.rb)) however seems a bit clunky. I suppose this is something Prometheus would help us with - with its focus on pulling metrics - so it might not be that we have a good alternative until then and it doesn't seem hugely pressing.

Reply:
Added a note about this